test/jobprogress.test.ts

```diff
@@ -0,0 +1,198 @@
+ *-----------------------------------------------------------------------------------------------*/
+
+import * as chai from 'chai';
+import * as sinon from 'sinon';
+import * as sinonChai from 'sinon-chai';
+import * as vscode from 'vscode';
+import { Protocol } from 'rsp-client';
+import { Stubs } from './stubs';
+import { JobProgress } from '../src/jobprogress';
+
+const expect = chai.expect;
+chai.use(sinonChai);
+
+class CancellationStub implements vscode.CancellationToken {
+  isCancellationRequested: boolean;
+  onCancellationRequested: vscode.Event<any> =
+    function() { return { dispose() { } }; };
+  }
+class ProgressStub implements vscode.Progress<{message: string, increment: number}> {
+  report(value: { message: string, increment: number }): void {
+  }
+}
+
+suite('Job Progress', () => {
+    let sandbox: sinon.SinonSandbox;
+    let stubs: Stubs;
+    const job: Protocol.JobHandle = {
+      name: 'papa smurf',
+      id: 'papa.smurf.id'
+    };
+    const otherJob: Protocol.JobHandle = {
+      name: 'smurfette',
+      id: 'smurfette.id'
+    };
+    let cancellationStub: vscode.CancellationToken;
+    let progressStub: vscode.Progress<{message: string, increment: number}>;
+    let progressStubReport: sinon.SinonSpy<[{ message: string; increment: number; }], void>;
+    let withProgressFake: (
+      options: vscode.ProgressOptions,
+      task: (progress: vscode.Progress<{ message: string; increment: number }>, token: vscode.CancellationToken) => Thenable<{}>) => Thenable<{}>;
+    let withProgressFakeSpy;
+
+    setup(() => {
+        sandbox = sinon.createSandbox();
+        stubs = new Stubs(sandbox);
+
+        stubs.incoming.onJobAdded = sandbox.stub();
+        stubs.incoming.onJobRemoved = sandbox.stub();
+        stubs.incoming.onJobChanged = sandbox.stub();
+
+        cancellationStub = new CancellationStub();
+        progressStub = new ProgressStub();
+        progressStubReport = sandbox.spy(progressStub, 'report');
+        withProgressFake = (
+          options: vscode.ProgressOptions,
+          task: (progress: vscode.Progress<{ message: string; increment: number }>, token: vscode.CancellationToken) => Thenable<{}>) => {
+              return task(progressStub, cancellationStub);
+        };
+        withProgressFakeSpy = sandbox.stub(vscode.window, 'withProgress').callsFake(withProgressFake);
+      });
+
+    teardown(() => {
+        sandbox.restore();
+    });
+
+    test('JobProgress registers onJobAdded listener', () => {
+      JobProgress.create(stubs.client);
+      expect(stubs.incoming.onJobAdded).calledOnce;
+    });
+
+    test('onJobAdded notification should open window with progress', () => {
+      // given
+      JobProgress.create(stubs.client);
+
+      // when
+      callOnJobAddedListenerWith(job, stubs.incoming.onJobAdded);
+
+      // then
+      expect(stubs.incoming.onJobAdded).calledOnce;
```

Comment:
you can extend the client call expectations to also verify the parameters
```calledOnceWith(job)```

Reply:
@jrichter1 you're right in general, but here onJobAdded is called with the listener to get informed of a new job. The assertion should make sure that jobProgress registered its (internal) listener, not a specific one.

Reply:
I admit, this line is about the worst example I could have picked, but I hope you get the idea :)